django/db/models/fields/json.py

```diff
@@ -0,0 +1,111 @@
+import json
+
+from django import forms
+from django.core import exceptions
+from django.db.utils import NotSupportedError
+from django.utils.translation import gettext_lazy as _
+
+from . import Field
+from .mixins import CheckFieldDefaultMixin
+
+
+def convert_jsonfield_value(value, expression, connection):
+    if connection.vendor == 'oracle':
+        return value.read()
+    return value
+
+
+class JSONField(CheckFieldDefaultMixin, Field):
+    description = _('A JSON object')
+    default_error_messages = {
+        'invalid': _("Value must be valid JSON."),
+    }
+    _default_hint = ('dict', '{}')
+
+    def __init__(self, encoder=None, decoder=None, default=dict, *args, **kwargs):
+        self.encoder, self.decoder = encoder, decoder
+        super().__init__(default=default, *args, **kwargs)
+
+    def _check_json_support(self, connection):
+        if not connection.features.supports_json:
+            raise NotSupportedError(_('This database backend does not support JSONField.'))
+
+    def db_check(self, connection):
+        data = self.db_type_parameters(connection)
+        if connection.vendor in ('mysql', 'sqlite'):
+            return 'JSON_VALID(%(qn_column)s)' % data
+        elif connection.vendor == 'oracle':
+            return '%(qn_column)s IS JSON' % data
+        return super().db_check(connection)
+
+    def db_type(self, connection):
+        self._check_json_support(connection)
+        if connection.vendor == 'postgresql':
+            return 'jsonb'
+        elif connection.vendor == 'oracle':
+            return 'clob'
+        elif connection.vendor == 'sqlite':
+            return 'text'
+        return 'json'
```

Comment:
Please move data types to `DatabaseWrapper.data_types` dicts for builtin backends.